### PR TITLE
[SEMI-MODULAR] Customization

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
-New HUD for character customization
-Re-write of the mutant bodyparts and features, addition of all the custom accessories from our base. The mutant bodyparts hold their own custom color key, which by default is either primary/secondary/tertiary/skin, depending on the accessory, but the user can change it

-Added I think all the accessories and customization that we've had previously

-Genitals are added, along with toggles to switch underwear or genital visibility (underwear will obstruct visibility of genitals)

-Mutant bodyparts are now very easily integratable with organs, and properly hold all the nessecary information and you can freely cut people's tails and transplant them etc.

-Addition of fully customizable per-limb body markings - They render just when bodyparts do AND I've removed the keyed limb cache for humans because it was doing us more harm than good, considering the vast customization and color options etc. Limb cache still is a thing for other carbons

-Addition of lots of flavor texts, OOC prefs, records, background informations. Records dont show up in consoles yet.

-Additional of loadout (and also the items added for the sake of the loadout). Loadout items can hold a piece of data and then manipulate it per-datum basis, which allows for color option of loadout items

-Added special render cases and sprites for taurs, digitigrades, voxes, and snouts

-Added about all the missing species I think

KNOWN ISSUES:
-Lack of a clip mask for taur clothing variants
-Missing sprites/wrongly set digi styles on clothing
-Missing/wrong sprites on undergarments

~~-Customizing a lizard can lead to some runtimes, possibly randomization giving you a mutant bodypart of a wrong key or something.~~ Fixed, this was due to TG doing some special handling of spines for the tail, for which I have another more widely purposeful system

## Changelog
:cl:

/:cl:

### CODE REVIEW

**Comment by Gandalf2k15** [comment]
Damage overlays appear to be broken, here is a picture of someone that has been burnt to a crisp: 
image

It looks like things are being reverted to their base colours.

Update: Brute is fine.

**Comment by Gandalf2k15** [comment]
Socks and underwear do not render for digitigrade legs.

**Comment by Gandalf2k15** [comment]
Runtime with exploding husky, 
`proc name: ex act (/mob/living/carbon/human/ex_act)
src: Fenton Werry (/mob/living/carbon/human)
src.loc: null
call stack:
Fenton Werry (/mob/living/carbon/human): ex act(3, null, null)
Explosions (/datum/controller/subsystem/explosions): fire(0)
Explosions (/datum/controller/subsystem/explosions): ignite(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): Loop()
Master (/datum/controller/master): StartProcessing(0)`

in

```
/mob/living/carbon/human/ex_act(severity, target, origin)
	if(TRAIT_BOMBIMMUNE in dna.species.species_traits)
		return
	..()
```

Cannot read null.species()

**Comment by Gandalf2k15** [comment]
Gibbed penis, testicles and tail do not show appropriate colour and icon

**Comment by Gandalf2k15** [comment]
A couple of minor issues with markings and where the neck meets the head:
With husky markings side view:
image
Front view: 
image

Without:
image

**Comment by Azarak** [comment]
> 
> 
> Damage overlays appear to be broken, here is a picture of someone that has been burnt to a crisp:
> image
> 
> It looks like things are being reverted to their base colours.
> 
> Update: Brute is fine.

Oh I know why, that's cause of husking. I'll make husking properly apply the gray color to everything

**Comment by Azarak** [comment]
> 
> 
> A couple of minor issues with markings and where the neck meets the head:
> With husky markings side view:
> image
> Front view:
> image
> 
> Without:
> image

It's a different rendering order than what was originally, though it isn't an issue with the order, it's an issue with the sprite, which should be adjusted, though I dont want to touch on small sprites changes in this PR

**Comment by Azarak** [comment]
> 
> 
> Runtime with exploding husky,
> `proc name: ex act (/mob/living/carbon/human/ex_act) src: Fenton Werry (/mob/living/carbon/human) src.loc: null call stack: Fenton Werry (/mob/living/carbon/human): ex act(3, null, null) Explosions (/datum/controller/subsystem/explosions): fire(0) Explosions (/datum/controller/subsystem/explosions): ignite(0) Master (/datum/controller/master): RunQueue() Master (/datum/controller/master): Loop() Master (/datum/controller/master): StartProcessing(0)`
> 
> in
> 
> ```
> /mob/living/carbon/human/ex_act(severity, target, origin)
> 	if(TRAIT_BOMBIMMUNE in dna.species.species_traits)
> 		return
> 	..()
> ```
> 
> Cannot read null.species()

seems unrelated, I don't touch removing species

**Comment by Azarak** [comment]
> 
> 
> Socks and underwear do not render for digitigrade legs.

Only an issue with some of them, I'd rather have people assemble a list of broken sprites, in a playtest, so I can fix them all up in a batch, for now ignoring issue

**Comment by Azarak** [comment]
> 
> 
> Damage overlays appear to be broken, here is a picture of someone that has been burnt to a crisp:
> image
> 
> It looks like things are being reverted to their base colours.
> 
> Update: Brute is fine.

Fixed

**Comment by Gandalf2k15** [comment]
List of broken under sprites for digitigrade:
**Underwear:**
- Men's Briefs
- Men's Striped Boxer
- Men's Midway Boxer
- Men's Long Johns
- Men's Hearts Boxer

Shouldn't be in underwear:
- Normal - Christmas
- Normal - Red Candy Cane
- Normal - Green Candy Cane

**Undershirt:**
None

**Socks:**

- _Knee-high:_
- black
- white
- _Normal:_
- black
- white
- _Short:_
- black
- white
- _Thigh-high:_
- black
- white
-

**Comment by Gandalf2k15** [comment]
You cannot hide genitals via underwear

**Comment by Gandalf2k15** [comment]
List of missing anthro sprites(will update as I find more):
**Jumpsuits:**
- /obj/item/clothing/under/rank/security/head_of_security
- /obj/item/clothing/under/rank/cargo/tech
- /obj/item/clothing/under/rank/medical/paramedic (has incorrect sprite?)

**Coats(don't need anthro variants):**
- /obj/item/clothing/suit/det_suit
- /obj/item/clothing/suit/toggle/labcoat/paramedic

**Comment by Gandalf2k15** [comment]
The balaclava hides ears, maybe need a setting for masks that specifically don't hide ears.

**Comment by Gandalf2k15** [comment]
Switching preview "loadout" from "naked" does not show your undergarments until you click "job"

**Comment by Azarak** [comment]
> 
> 
> Switching preview "loadout" from "naked" does not show your undergarments until you click "job"

Intended. Naked is literally naked

File: modular_skyrat/code/modules/mob/living/carbon/carbon_update_icons.dm [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		var/used_style = NONE
+		if(dna?.species.id == "vox")
+			used_style = STYLE_VOX
+		else if(dna?.species.mutant_bodyparts["snout"])
+			var/datum/sprite_accessory/snouts/S = GLOB.sprite_accessories["snout"][dna.species.mutant_bodyparts["snout"][MUTANT_INDEX_NAME]]
+			if(S.use_muzzled_sprites && head.mutant_variants & STYLE_MUZZLE)
+				used_style = STYLE_MUZZLE
+		switch(used_style)
+			if(STYLE_MUZZLE)
+				desired_icon = head.worn_icon_muzzled || 'modular_skyrat/icons/mob/clothing/head_muzzled.dmi'
```
Comment (Gandalf2k15) [inline modular_skyrat/code/modules/mob/living/carbon/carbon_update_icons.dm]: This should be in the master_files section.

Comment (Azarak) [reply modular_skyrat/code/modules/mob/living/carbon/carbon_update_icons.dm]: Done

File: code/modules/mob/living/carbon/human/emote.dm [hunk 2]
```diff
@@ -104,7 +104,7 @@
 	var/mob/living/carbon/human/H = user
 	if(!istype(H) || !H.dna || !H.dna.species || !H.dna.species.can_wag_tail(H))
 		return
-	if(!H.dna.species.is_wagging_tail())
+	if(!H.dna.species.is_wagging_tail(H)) //SKYRAT EDIT CHANGE - CUSTOMIZATION
```
[the remark this answers is not in the document]
Comment (Gandalf2k15) [reply code/modules/mob/living/carbon/human/emote.dm:108]:
```suggestion
	//if(!H.dna.species.is_wagging_tail()) - ORIGINAL
	if(!H.dna.species.is_wagging_tail(H)) //SKYRAT EDIT CHANGE - CUSTOMIZATION
```

File: code/modules/mob/living/carbon/human/emote.dm [hunk 3]
```diff
@@ -120,7 +120,7 @@
 	var/mob/living/carbon/human/H = user
 	if(!H.dna || !H.dna.species)
 		return
-	if(H.dna.species.is_wagging_tail())
+	if(H.dna.species.is_wagging_tail(H)) //SKYRAT EDIT CHANGE - CUSTOMIZATION
```
Comment (Gandalf2k15) [inline code/modules/mob/living/carbon/human/emote.dm:125]: Ditto.
```suggestion
	//if(H.dna.species.is_wagging_tail()) - ORIGINAL
	if(H.dna.species.is_wagging_tail(H)) //SKYRAT EDIT CHANGE - CUSTOMIZATION
```

File: code/modules/mob/living/carbon/human/human_update_icons.dm [hunk 4]
```diff
@@ -559,8 +579,8 @@ generate/load female uniform sprites matching all previously decided variables
 		. += "-coloured-[skin_tone]"
 	else if(dna.species.fixed_mut_color)
 		. += "-coloured-[dna.species.fixed_mut_color]"
-	else if(dna.features["mcolor"])
-		. += "-coloured-[dna.features["mcolor"]]"
+	else if(MUTCOLORS in dna.species.species_traits) //SKYRAT EDIT CHANGE - CUSTOMIZATION
+		. += "-coloured-[dna.features["mcolor"]]" //SKYRAT EDIT CHANGE - CUSTOMIZATION
```
Comment (Gandalf2k15) [inline code/modules/mob/living/carbon/human/human_update_icons.dm:585]: Original keep.
```suggestion
	//else if(dna.features["mcolor"]) - ORIGINAL
	//	. += "-coloured-[dna.features["mcolor"]]" - ORIGINAL
	else if(MUTCOLORS in dna.species.species_traits) //SKYRAT EDIT CHANGE - CUSTOMIZATION
		. += "-coloured-[dna.features["mcolor"]]" //SKYRAT EDIT CHANGE - CUSTOMIZATION
```

File: code/modules/mob/living/carbon/human/species.dm [hunk 5]
```diff
@@ -57,7 +57,7 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	/// Default mutant bodyparts for this species, like horns. Don't forget to set one for every mutant bodypart you allow this species to have.
 	var/list/default_features = list()
 	/// Visible CURRENT bodyparts that are unique to a species. DO NOT USE THIS AS A LIST OF ALL POSSIBLE BODYPARTS AS IT WILL FUCK SHIT UP! Changes to this list for non-species specific bodyparts (ie cat ears and tails) should be assigned at organ level if possible. Layer hiding is handled by [datum/species/handle_mutant_bodyparts()] below.
-	var/list/mutant_bodyparts = list()
+	var/list/list/mutant_bodyparts = list() //SKYRAT EDIT CHANGE - CUSTOMIZATIOn
```
Comment (Gandalf2k15) [inline code/modules/mob/living/carbon/human/species.dm:61]:
```suggestion
	//var/list/mutant_bodyparts = list() - ORIGINAL
	var/list/list/mutant_bodyparts = list() //SKYRAT EDIT CHANGE - CUSTOMIZATIOn
```

**Comment by Gandalf2k15** [comment]
Why are these files in master_files? 
image

**Review by Gandalf2k15 — CHANGES_REQUESTED** [review]
Final check run #1 complete. Final check run #2 required before merging. Independent inspectorate.

Please ensure you update the readme.md file in the master_files section with your additions.

File: code/_globalvars/lists/flavor_misc.dm [hunk 6]
```diff
@@ -17,6 +17,8 @@ GLOBAL_LIST_EMPTY(undershirt_f)	 //stores only undershirt name
 	//Socks
 GLOBAL_LIST_EMPTY(socks_list)		//stores /datum/sprite_accessory/socks indexed by name
 	//Lizard Bits (all datum lists indexed by name)
+//SKYRAT EDIT REMOVAL - CUSTOMIZATION
```
[the remark this answers is not in the document]
Comment (Azarak) [reply code/_globalvars/lists/flavor_misc.dm:20]: Nope, removed straight up

**Pull request merged**